File: CS50_week2/week2pset/readability/readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int calculateReadingLevel(string text);
int howManyWords(string text);
int howManyLetters(string text);
int howManySentences(string text);
int outputGrade(int index);

int main(void)
{
    string text = get_string("Text: ");
    int level = calculateReadingLevel(text);
    int output = outputGrade(level);

    if (output == 2)
    {
        printf("Before Grade 1\n");
    }
    else if (output == 1)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", level);
    }
}

int calculateReadingLevel(string text)
{
    int letters = howManyLetters(text);
    int words = howManyWords(text);
    int sentences = howManySentences(text);

    float L = letters * (100.0 / words);
    float S = sentences * (100.0 / words);
    float index = (0.0588 * L) - (0.296 * S) - 15.8;
    return round(index);
}

int howManyWords(string text)
{
    int spaces = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isblank(text[i]))
        {
            spaces++;
        }
    }
    return spaces + 1;
}

int howManyLetters(string text)
{
    int letters = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
    }
    return letters;
}

int howManySentences(string text)
{
    int sentences = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == '?' || text[i] == '!' || text[i] == '.')
        {
            sentences++;
        }
    }
    return sentences;
}

int outputGrade(int index)
{
    if (index < 1)
    {
        return 2;
    }
    else if (index >= 16)
    {
        return 1;
    }
    else
    {
        return 0;
    }
}

File: CS50_week6/week6pset/sentimental-readability/readability.py
from cs50 import get_string


def main():
    text = get_string("Text: ")
    index = calculateColemanLiauIndex(text)
    print(gradeLevel(index))


def calculateColemanLiauIndex(text):
    letters = numLetters(text)
    words = numWords(text)
    sentences = numSentences(text)

    L = letters * (100 / words)
    S = sentences * (100 / words)
    index = (0.0588 * L) - (0.296 * S) - 15.8
    return round(index)


def numLetters(text):
    numLetters = 0
    for i in range(len(text)):
        if text[i].isalpha():
            numLetters += 1

    return numLetters


def numWords(text):
    numSpaces = 0
    for i in range(len(text)):
        if text[i] == " ":
            numSpaces += 1
    numWords = numSpaces + 1

    return numWords


def numSentences(text):
    numSentences = 0
    for i in range(len(text)):
        if text[i] == '?' or text[i] == '!' or text[i] == '.':
            numSentences += 1

    return numSentences


def gradeLevel(index):
    if index < 1:
        return "Before Grade 1"
    elif index >= 16:
        return "Grade 16+"
    else:
        return f"Grade {index}"


main()
